pd-client/client.go

```diff
@@ -14,106 +14,401 @@
+				continue
+			}
+			c.clusterID = clusterID
+			return nil
+		}
+
+		time.Sleep(time.Second)
+	}
+
+	return errors.Trace(errFailInitClusterID)
 }
 
-func (c *client) GetClusterID() uint64 {
-	return c.worker.clusterID
+func (c *client) updateLeader() error {
+	for _, u := range c.urls {
+		// TODO: Use gRPC instead.
+		client, err := apiutil.NewClient(u, pdTimeout)
+		if err != nil {
+			continue
+		}
+		leader, err := client.GetLeader()
+		if err != nil {
+			continue
+		}
+		if err = c.switchLeader(leader.GetAddr()); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	}
+	return errors.Errorf("failed to get leader from %v", c.urls)
 }
 
-func (c *client) GetTS() (int64, int64, error) {
-	req := &tsoRequest{
-		done: make(chan error, 1),
+func (c *client) switchLeader(addr string) error {
+	c.connMu.RLock()
+	if c.connMu.leader == addr {
+		c.connMu.RUnlock()
+		return nil
+	}
+
+	log.Infof("[pd] leader switches to: %v, previous: %v", addr, c.connMu.leader)
+	if _, ok := c.connMu.clientConns[addr]; ok {
+		c.connMu.RUnlock()
+		c.connMu.Lock()
+		c.connMu.leader = addr
+		c.connMu.Unlock()
+		return nil
+	}
+	c.connMu.RUnlock()
+
+	cc, err := grpc.Dial(addr, grpc.WithDialer(func(addr string, d time.Duration) (net.Conn, error) {
+		u, err := url.Parse(addr)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		// For tests.
+		if u.Scheme == "unix" || u.Scheme == "unixs" {
+			return net.DialTimeout(u.Scheme, u.Host, d)
+		}
+		return net.DialTimeout("tcp", u.Host, d)
+	}), grpc.WithInsecure()) // TODO: Support HTTPS.
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	c.connMu.Lock()
+	defer c.connMu.Unlock()
+	if _, ok := c.connMu.clientConns[addr]; ok {
+		cc.Close()
+	} else {
+		c.connMu.clientConns[addr] = cc
+	}
+	c.connMu.leader = addr
+	return nil
+}
+
+func (c *client) leaderLoop() {
+	defer c.wg.Done()
+
+	for {
+		select {
+		case <-c.checkLeaderCh:
+		case <-time.After(time.Minute):
+		case <-c.ctx.Done():
+			return
+		}
+
+		if err := c.updateLeader(); err != nil {
+			log.Errorf("[pd] failed updateLeader: %v", err)
+		}
+	}
+}
+
+func (c *client) tsLoop() {
+	defer c.wg.Done()
+
+	var requests []*tsoRequest
+	var stream pdpb2.PD_TsoClient
+	var cancel context.CancelFunc
+
+	for {
+		var err error
+
+		if stream == nil {
+			var ctx context.Context
+			ctx, cancel = context.WithCancel(c.ctx)
+			stream, err = c.leaderClient().Tso(ctx)
+			if err != nil {
+				log.Errorf("[pd] create tso stream error: %v", err)
+				time.Sleep(time.Second)
+				continue
+			}
+		}
+
+		select {
+		case first := <-c.tsoRequests:
+			requests = append(requests, first)
+			pending := len(c.tsoRequests)
+			for i := 0; i < pending; i++ {
+				requests = append(requests, <-c.tsoRequests)
+			}
+			done := make(chan struct{})
+			go func(cancel context.CancelFunc) {
+				select {
+				case <-time.After(pdTimeout):
+					cancel()
+				case <-done:
+				}
+			}(cancel)
+			err = c.processTSORequests(stream, requests)
+			close(done)
```

Comment:
when we close the done here, another coroutine may timeout and call cancel at the same time, so we should check ctx.Err() below.